src/Agent.Worker/AsyncCommandContext.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker
+{
+    [ServiceLocator(Default = typeof(AsyncCommandContext))]
+    public interface IAsyncCommandContext : IAgentService
+    {
+        string Name { get; }
+        Task Task { get; set; }
+        void InitializeCommandContext(IExecutionContext context, string name);
+        void Output(string message);
+        Task WaitAsync();
+    }
+
+    public class AsyncCommandContext : AgentService, IAsyncCommandContext
+    {
+        private readonly TaskCompletionSource<int> _asyncCommandCompleted = new TaskCompletionSource<int>();
+        private IExecutionContext _executionContext;
+        private readonly ConcurrentQueue<string> _outputQueue = new ConcurrentQueue<string>();
+
+        public string Name { get; private set; }
+        public Task Task { get; set; }
+
+        public void InitializeCommandContext(IExecutionContext context, string name)
+        {
+            _executionContext = context;
+            Name = name;
+        }
+
+        public void Output(string message)
+        {
+            _outputQueue.Enqueue(message);
+        }
+
+        public async Task WaitAsync()
+        {
+            Trace.Entering();
+
+            // start flushing output queue
+            Trace.Info("Start flush buffered output.");
+            Task flushOutput = FlushAsync();
+            try
+            {
+                // wait for the async command task
+                Trace.Info("Wait till async command task to finish.");
+                await Task;
+            }
+            finally
+            {
+                // fire signal to finish flushing output queue.
+                Trace.Info("Finsh flush buffered output.");
+                _asyncCommandCompleted.TrySetResult(0);
+
+                // wait for all output flushed.
+                await flushOutput;
+            }
+        }
+
+        private async Task FlushAsync()
+        {
+            _executionContext.Section($"Async Command Start: {Name}");
+
+            string output;
+            while (!_asyncCommandCompleted.Task.IsCompleted)
```

Comment:
I should use this.Task.IsCompleted, will merge the logic, only have WaitAsync().